### PR TITLE
Engineering analysis: `cae:`, `pc cae fea|cfd`, and the IDE's FEA and CFD tabs

### PR DESCRIPTION
**Pull request by clairbee**
## Copilot Summary

Computer-aided engineering becomes the third thing PartCAD produces from a shape, beside a file another tool
opens (`export:`) and a picture of it (`render:`) — and it is built out of that machinery rather than beside it.

### `cae:` is an output section of the same shape

One subsection per file type, `path`/`package` naming the script, the same sandbox and the same meta-wrapper.
`Shape._run_implementation_async()` is the body all three sections now share, so a change to how an
implementation is run cannot reach two of them and miss the third.

It is deliberately **not** in `output.SECTIONS`: that tuple answers "which sections does a `pc export`/`pc render`
file type live in", and a `fea` left in there would be offered to `pc render -t` and would fall back to a render
implementation. It also has no built-in package — PartCAD ships no solver — so `output.builtin_project()` answers
`None` for it and everything downstream copes with a missing bottom layer.

### The part declares the boundary conditions

In a section named after the analysis, because they belong to the part and not to whoever analyses it: a bracket
is bolted down at the same holes whichever solver is asked.

```yaml
parts:
  bracket:
    type: build123d
    path: bracket.py
    fea:
      fix:
        - m3-screw          # every instance of this interface is held still
      load:
        hook: 5 kg          # every instance carries this
        rail:
          left: 30 N        # or one named instance at a time
          right: 30 N
```

A `load` is a **force**. Write a number and a unit — `n`/`nm`/`mn`/`kn`/`newton` for force,
`mg`/`g`/`kg`/`ton`/`tonne`/`lb`/`pound` for mass, matched case-insensitively, with or without a space and with
or without a plural `s` — or a bare number, which is a mass in kilograms weighed into newtons at 9.8 N/kg.
`partcad/cae.py` parses all of it, converts the units, and attaches each condition to the ports
`render_overlay.collect_async()` already finds — so `pc render --with-ports` draws exactly what the solver was
told. It imports nothing from `partcad`, which is what lets it be tested without a sandbox.

`cfd:` takes the same two keys with the same meaning: what is held still, and what force the flow puts on it.

### Findings are the other half of the answer

The implementation reports `findings` beside the model it wrote: a JSON array that `pc cae` prints, that
`pc test` fails on, and that the IDE lists under the model. The model goes to `<part>.<analysis>.<extension>`
and is saved as it stands; which format that is — 2D or 3D — is the implementation's decision, which is why
`extension:` is required of a `cae:` type and PartCAD does not guess one.

### What this adds

- **`pc cae fea|cfd <part>`**, a daemon command, exiting non-zero when there is a finding so it can be used as
  a gate. Both are one operation with the analysis as an argument; `partcad_cli/click/analysis.py` holds the
  options and the body the way `viewport.py` holds the three `pc render` and `pc adhoc render` share.
- **`pc test` gains an `fea` and a `cfd` check**, applicable only to a part that declares the matching section —
  a `pc test -r` must not start a solver for every bolt in a package tree, and declaring `fea:` is how a user
  asks.
- **`caeFeaImplementation`/`caeCfdImplementation`** in the user configuration name who runs each analysis, as
  `<package>:<file type>`, defaulting to `//pub/feature/cae/calculix:fea` and `:cfd`.
  `-i`/`--implementation` overrides one run, and a part may name its own with `implementation:` (below).
- **The VS Code viewer gains an FEA and a CFD tab for a part**, backed by `cae.analyze`. They are the one pair
  in that panel that acts rather than asks: selecting one runs the solver. The implementation is a field over
  the model, pre-filled from `cae.defaults` and filled in even when the run failed — which is when a user most
  needs to see what was tried. The model is drawn by its extension (glb/gltf/stl on an orbit camera,
  png/jpg/svg and friends as an image that pans and zooms), and any findings take the bottom fifth of the pane,
  which a clean part gets back in full.

### The implementation itself

[`partcad/partcad-cae-calculix`](https://github.com/partcad/partcad-cae-calculix), registered in the public
index by [partcad/partcad-index#10](https://github.com/partcad/partcad-index/pull/10) (merged). It is a package
like any other and is not part of this repository.

**It has now been run end to end against CalculiX 2.23**, and the cases below are what found the four defects
stacked in front of the solver — see "What a real solver found" below. `fea` is validated; `cfd` is not, and
that package's `README.md` says so before it says anything else.

### Cases with a known answer

`examples/feature_cae` is a pair of textbook problems whose answer is a formula, so what a solver returns can be
compared against arithmetic instead of against expectations. The unit tests show the code does what its author
meant; these are the only thing that can show that what the author meant is what CalculiX expects.

**`cantilever`** — 100×10×10 mm of steel, clamped at one end, 100 N at the other. `δ = PL³/3EI` = 0.2000 mm and
`σ = 6PL/bh²` = 60.0 MPa exactly, plus 0.8% for shear, so 0.2016 mm from a perfectly clamped 3D model.
`self_weight` is pinned off because the closed form has no gravity in it and the implementation's default does.

It is *not* expected to read 0.2016 mm, and the README says why: `fix:` names a port, a port becomes a ball of
`port_radius`, and the 7.5 mm ball that covers the end face clamps part of the length with it. The README
predicted **0.16–0.18 mm** on that reasoning. **Measured: 0.1655 mm**, peak von Mises 58.28 MPa — inside the
window, and mesh-converged (0.1650 / 0.1655 / 0.1631 mm across 434, 2375 and 6468 elements). The clamp was
measured reaching 6.8 mm in from each end, for a free span of 86.4 mm, which is exactly the cube-law arithmetic
the README describes. So the case did its job: **getting within a few percent needs `fix:` to clamp a face**,
and that is now a measurement rather than a prediction.

**`pipe`** — the bore of a 10×100 mm tube driven by 5 mN. Hagen–Poiseuille gives `u_max` = 0.03979 m/s, at
`Re` = 1.79 with an entrance length of 1.1% of the pipe, so the regime is checked rather than assumed.

**This one produced no measurement at all, and not for the reason predicted.** The README had it disagreeing
because `fix:` cannot say "the lateral surface", so the no-slip wall is 20 port-balls with gaps. That prediction
was untested and is wrong: the CFD solve *diverges*, to a dead field of order 1e-16 m/s. At least one cause is
structural and is PartCAD's rather than the implementation's — **`cfd:` can name walls and an inlet and has no
way to name an outlet**, so the incompressible problem is posed with no downstream pressure reference. The
README now says that, and is labelled as having been wrong.

Neither part renders anything — no `render:` section — so `pc render -r` produces nothing new and the
checked-in-examples diff has nothing to compare.

### A machine with no solver is no longer a failing part

Adding the example above made an earlier decision obviously wrong. `pc test`'s `fea`/`cfd` checks used to *fail*
a part whose analysis could not be run at all. PartCAD ships no solver and `caeFeaImplementation` names a
third-party package needing a native binary, so under that rule one part declaring `fea:` broke `pc test` for
every contributor and every CI runner without CalculiX — including this repository's own examples. A check that
cannot be satisfied by a default install on a default machine is not a check.

It is now reported and passed over, the way `cam` passes a part that is `manufacturable: false`. Loudly, and as
a warning rather than an error, because it says something about the machine and nothing about the part. A
malformed section still fails, and an analysis that ran and reported findings still fails. The cost is stated in
the docstring: nothing there can tell a solver that is absent from one that crashed, so a crashing solver is
skipped too — `pc cae fea` is what reports that as the error it is. **The validation below shows that cost is
real**: `pc test -f cfd` is green against a CFD implementation that does not work at all.

That skip is now narrower than it was — a plugin that does not *resolve* is a failure, see below. Only the case
this section is about, a resolved plugin with no solver binary on this machine, still passes over.

### What a real solver found

The cases above were run against **CalculiX 2.23** (conda-forge, macOS arm64) — the first execution of any of
this. Four defects sat in front of the solver, all four in
[`partcad-cae-calculix`](https://github.com/partcad/partcad-cae-calculix) rather than here, and all four are
fixed and pushed there (`4a6c1c1`, `41719a7`):

1. **`trimesh==4.9.1` does not exist** — PyPI goes 4.9.0 → 4.10.0, so provisioning the sandbox failed outright.
   (`numpy==2.4.0` is also yanked; both repinned.)
2. **`import calculix_common` could not resolve** — `runpy.run_path()` does not put the script's directory on
   `sys.path`. PartCAD's own `builtin/render/render_dxf.py` shows the two lines needed.
3. **gmsh's TET10 is not CalculiX's C3D10** — the last two mid-side nodes are swapped, so `ccx` rejected *every
   element in every mesh* with a nonpositive jacobian, and the empty `.frd` was then reported as "the analysis
   did not converge". `mesh_order: 2` is the default, so this was the whole of FEA.
4. **Two node sets were both named `PORT0`** — CalculiX merges them silently, so the encastre held the loaded
   end too and the load pushed on nodes that could not move. **The answer was `0.0000 mm`**: a valid deck, a
   happy solver, and a number a reader might believe. This is the one to prioritise, and it is why a case with
   a known answer is worth having at all.

Also fixed there: the CFD deck was missing four cards CalculiX names one refusal at a time, read its result
blocks under structural names (`VELO`/`PRESSURE` rather than `V3DF`/`PS3DF`), and wrote one result block per
increment against `INCF=1000000` — one observed run reached a 5 GB `.frd` and was still growing. The deck is
accepted now; it still does not converge, and that README leads with it.

**The index prerequisite is worked around** rather than outstanding. `//pub/feature/cae/calculix` is registered
on the index's `devel` branch and has not reached `main`, which is what `examples/partcad.yaml` and every
`pc init` project pin — so the default `caeFeaImplementation` resolves to nothing. Repinning the examples tree
at the index's `devel` would fix it and would also rename the index root `/pub` → `//pub`, which reaches every
example in the tree, so instead `examples/feature_cae` depends on `partcad-cae-calculix` directly and each part
names it with `implementation:`. Both the `dependencies:` entry and the two `implementation:` lines are
temporary, carry a comment saying to delete them once the index carries `feature/cae`, and are explained in the
example's README.

### After review: one lock, a part that names its solver, and a plugin that must exist

Three changes asked for on this PR, and the two things CI then found.

**One lock for a shape.** `Shape.locked()` is a re-entrant async context manager taking the *same* pair the
instantiation path already takes — the `threading.RLock` plus the per-thread, per-loop `asyncio.Lock`. It now
also covers producing the shape's files, because the output path is derived from the shape: `analyze_async`
holds it across clear-run-verify, and `_run_implementation_async` takes it too, so `export:` and `render:` are
covered by the same change. Re-entrancy is load-bearing rather than decorative — the operations nest
(`analyze_async` → `get_wrapped` → `_run_implementation_async`) and a bare `asyncio.Lock` would deadlock — so
the owning task is remembered and passes straight through. The cost is stated in the docstring: two *different*
outputs of one shape no longer overlap, even though they write different paths. What it does **not** make the
returned path is a snapshot; a comment at the lock says so, and says why that is the same contract `pc render`
and `pc export` have.

**A part may name the solver it was written against**, with `implementation:` in its `fea:`/`cfd:` section,
ranking below `-i` and above the user configuration. A *relative* name here resolves against the package that
declared it — not against the current package, the way a name a user types does. That distinction is not
cosmetic: `pc test -r` over a tree runs with the tree's root current and every part one or more packages below
it, so resolving a package's own declaration against the current package makes it mean different things
depending on which directory the command was run from. `Examples (PartCAD)` went red on every leg until that
was fixed, and three tests now pin it on a package tree shaped like the one CI walks.

**A missing or misconfigured plugin fails `pc test`.** The implementation and its options are resolved before
anything runs, so a package that is not a dependency, did not load, or declares no such file type is a failure
with the sentence naming what did not resolve — that is wrong on every machine, and no amount of installing a
solver fixes it. Only the fourth outcome still skips: the plugin resolved and there is no solver binary here.

**What that then exposed**, on the two `ubuntu-24.04-arm` legs and nowhere else: `gmsh` publishes four wheels
per release — macOS x86_64, macOS arm64, manylinux x86_64 and win_amd64 — and **no linux aarch64 wheel and no
sdist**, in every release from 4.12 to 4.15. So pip failed hard, PartCAD logged that as an error, and a logged
error latches the flag that makes `pc` exit non-zero *even though* `CaeTest` had already caught the exception
and skipped the part. Fixed in the implementation package
([`971dc70`](https://github.com/partcad/partcad-cae-calculix/commit/971dc70)) with a marker on the requirement
— the same technique `sandbox_versions.py` already uses for the numba/Intel-macOS gap: pip then skips gmsh and
exits clean, and the missing module is reported as `SolverMissing`, which is the path a missing `ccx` already
takes. Both arm legs are green.

### Testing

- `tests/partcad/unit/test_cae.py` — units (a load written `5` and one written `5 kg` are the same force;
  `2 lb` is pounds and not two of something ending in `b`; `moons` is not some number of newtons), `fix:`/`load:`
  parsing and every way it can be malformed, attaching conditions to ports, findings.
- `tests/partcad/unit/test_cae_output.py` — that `cae:` is not an output format and does not fall back to one,
  the file naming, resolving an implementation (including a relative one, from a tree root), the meta-wrapper
  carrying findings back, and every branch of the `pc test` checks including the skip above.
- `tests/partcad/unit/test_shape_lock.py` — one holder at a time, re-entrancy within a task, a second task still
  waiting, release on exception, two runs not taking each other's model, and per-loop locking across threads.
- `tests/partcad_cli/unit/test_cae.py` — what `pc cae fea|cfd` sends, the absolute `--output-dir`, and the exit
  status: a finding is the answer "no", so the command stops a `&&` chain the way `pc test` does.
- `tests/partcad_service_json_rpc/test_operations.py` — `cae.analyze`/`cae.defaults` through the fake session.
- The whole of `tests/partcad/unit` was run with and without the resolution fix: the same 36 tests fail on both
  sides (all pre-existing, in a container with no CAD kernel) and 1886 → 1890 pass, which is exactly the four
  tests that fix added.
- 581 tests pass across the suites this touches. `src/partcad/cae.py`, `src/partcad/test/cae.py` and the CLI
  are at 100%; what is left uncovered in `shape.py` is the sandbox path, deliberately.
- The extension typechecks, lints, formats and builds clean.
- **End to end against CalculiX 2.23**, per the section above: `pc cae fea :cantilever` reads 0.1655 mm and
  `pc test -f fea` passes; `PARTCAD_CCX=/nonexistent` makes `pc cae fea` exit 1 with the error and `pc test -f
  fea` warn and pass, which is the new behaviour this PR introduces, confirmed on a real machine.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01NiUNXardYk3xmkDGjvUC6W

### CODE REVIEW

File: src/partcad_cli/click/analysis.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            "'caeCfdImplementation' user configuration option"
+        ),
+        type=str,
+        show_envvar=True,
+    ),
+    click.option(
+        "-O",
+        "--output-dir",
+        help="Write the model into the given directory instead of where the configuration puts it",
+        type=click.Path(exists=True, file_okay=False, dir_okay=True),
```
[the remark this answers is not in the document]
Comment (clairbee) [reply src/partcad_cli/click/analysis.py:48]: Not taking this one, because it would make `pc cae` the odd one out.

`--output-dir` and `-p/--create-dirs` sit side by side on `pc render` (`src/partcad_cli/click/commands/render.py:23,30-32`) and on `pc export` (`src/partcad_cli/click/commands/export.py:17,23-25`), both with `exists=True`. `pc cae` copies that pair deliberately: the three commands write artifacts the same way and a user who has learned the flags on one should not find them behaving differently on another.

You are right that the pair reads oddly — `--create-dirs` is about the directory structure created *under* the output directory, not about the output directory itself, and the flag name does not say so. But that is one decision about three commands, and changing it here alone would leave the CLI disagreeing with itself. Worth its own issue rather than a corner of this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/partcad_service_json_rpc/core/operations.py [hunk 2]
```diff
@@ -1889,6 +1889,105 @@ def assembly_guide(session, params):
...
+
+    shape = ctx.get_part(path)
+    if shape is None:
+        # Only a part is analysed. An assembly is a set of parts that each have
+        # their own boundary conditions, and a "load" on the whole of one says
+        # nothing about which of its members carries it.
+        raise JsonRpcError(USAGE_ERROR, "Part %s is not found" % path)
+
+    with pc.logging.Process(analysis.upper(), package, object_name):
+        ctx.option_create_dirs = bool(params.get("create_dirs", False))
```
[the remark this answers is not in the document]
Comment (clairbee) [reply src/partcad_service_json_rpc/core/operations.py:1957]: The race is real, but it is not this PR's and the fix does not belong here.

`ctx.option_create_dirs` is a shared mutable flag on the context, and `render_objects` sets it exactly the same way, with no lock and no save/restore (`src/partcad_service_json_rpc/core/operations.py:2338`). `cae_analyze` is one more operation doing what every file-writing operation on this context already does; making it the single one that serializes would not remove the race — a concurrent `render` would still overwrite the flag out from under it — it would only make the code look as if the problem had been dealt with.

The actual fix is to stop the flag being context state at all and pass the directory policy through the call, which touches `Context`, `ensure_dirs()` and every operation that sets it. That is a change worth making on its own, where it can be reviewed as one thing.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: ide/vscode/src/webview/cae.ts [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+            return;
+        }
+
+        if (MESH_TYPES.has(extension)) {
+            const view = new MeshView(this.model);
+            this.meshView = view;
+            void view.show(decodeBase64(data.content), extension).catch((error: unknown) => {
+                empty(this.model);
+                this.model.appendChild(el('p', 'error', `Failed to display the model: ${error}`));
+            });
```
[the remark this answers is not in the document]
Comment (clairbee) [reply ide/vscode/src/webview/cae.ts:470]: Fixed with a render generation: the model band counts its contents, `render()` captures the count it created, and a `show()` rejection that is not the current one returns without touching the DOM. The same count is what `MeshView.show()` reads before adding a parsed mesh to a scene that has since been disposed of.

No regression test, and not for want of trying: `ide/vscode` has no renderer-side test harness at all. Everything under `src/test/suite` runs in the extension host under `vscode-test`, where there is no `document`, no WebGL and no `three` — so a test for this would mean bringing in jsdom and a stub for the three.js pieces before the first assertion could be written. That is a harness for the whole `webview/` directory rather than a test for this line, and worth doing as its own change.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/partcad/cae.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+    for key, declared in (("fix", config.fixtures), ("load", config.loads)):
+        for name in declared:
+            if "%s:%s" % (key, name) in matched:
+                continue
+            if name in interfaces_seen:
+                wanted = declared[name] if key == "fix" else list(declared[name])
+                named = ", ".join(sorted(str(one) for one in wanted if one != EVERY_INSTANCE))
+                unmatched.append(
+                    (name, "no instance named %s" % named if named else "it names no instance of it")
+                )
```
[the remark this answers is not in the document]
Comment (clairbee) [reply src/partcad/cae.py]: Confirmed and fixed in `0977f21`.

You are right, and it is the worst shape a mistake can take here: the solver runs, holds one bolt of the two it was told to hold, and answers with a number that looks perfectly reasonable. `matched` now records `(key, interface, the instance the declaration named)` rather than the interface alone, and the unmatched report names the instances that landed nowhere.

Four regression tests, each of which fails on the old code — a `fix:` naming two instances where the object has one, the same for a per-instance `load:`, that `fix: [rail]` (every instance) is still satisfied by any one port, and that an interface-wide load and an instance's own load are each accounted for separately, since either is a declaration a user could have got wrong on its own.

The `shape.py` artifact check from the same review is in that commit too: `analyze_async` now refuses rather than returning a `filepath` nothing wrote, which is the check `_convert_to_serialized()` already makes of an exporter.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/partcad/shape.py [hunk 5]
```diff
@@ -1209,6 +1248,261 @@ def render(
...
+            result = await self._run_implementation_async(ctx, impl, script, request, final_filepath)
+
+        if result is None:
+            raise Exception("The '%s' implementation reported nothing: %s" % (format_name, script))
+        if not result.get("success", False):
+            raise Exception(
+                "%s failed for %s:%s: %s"
+                % (analysis.upper(), self.project_name, self.name, result.get("exception", "Unknown error"))
+            )
+        if not os.path.exists(final_filepath):
```
[the remark this answers is not in the document]
Comment (clairbee) [reply src/partcad/shape.py:1478]: Right, and the gap this leaves is worse than the one the last commit closed. Fixed in `a84d801`.

The path is stable — `<part>.<analysis>.<extension>`, beside the package — so an earlier run's file satisfies `os.path.exists`, and an implementation that reports success without writing gets last week's stresses handed back as the answer to today's load, with nothing anywhere to say they are not today's. A missing model is a refusal; a stale one is a wrong answer that looks like a right one.

I took the first of your two options: the path is removed before the implementation is asked, right after `ensure_dirs_for_file`. That is what makes the file's existence afterwards mean what it is read as meaning, and it leaves no staging path to reason about. The cost is that a failed run leaves no model behind at all — which seems right, given that the one it would have left is exactly the stale one this is about.

No regression test, for the same reason the check it repairs has none: reaching either means geometry, the port overlay, a materialized script and the sandbox, and mocking all four to exercise two lines would assert nothing about any of them. Both are recorded as uncovered in their commit messages.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/partcad/test/cae.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            return self.failed(shape, "%s", e)
+        except Exception as e:
+            # No verdict, so nothing to fail the part with. Loudly, because a
+            # silent pass here reads as "the part was checked" when nothing of
+            # the sort happened -- see the docstring.
+            pc_logging.warning(
+                "Test skipped: %s:%s: %s was not run: %s"
+                % (shape.project_name, shape.name, self.analysis.upper(), e)
+            )
+            return self.TEST_PASSED
```
[the remark this answers is not in the document]
Comment (clairbee) [reply src/partcad/test/cae.py:215]: Taken, in `c789802` — and thank you, the caching half of this is a real bug I had not seen. The validation run on a machine with CalculiX hit it in practice: its first attempt at the skip path returned exit 0 in 0.02 s from cache without exercising anything.

You are right about the mechanism. `CaeTest.cache_key_suffix()` folds in the boundary conditions, the resolved implementation and its options — everything about the *question* — and nothing about the machine, because a test cannot know what its implementation needs installed. So installing CalculiX changes no key and the remembered pass outlives its reason.

The fix is `Test.NOT_CACHEABLE`, a key a check may set on its `test_ctx` to mean "this verdict is about the machine"; `test_cached` then skips the write. `CaeTest` sets it on the skip branch and nowhere else, so an analysis that actually ran — pass or fail — stays cacheable, which is what the cache is for when running one costs a solver. Two things fell out of it: `test_cached` now copies `test_ctx` before handing it down, and `CaeTest.test` takes `test_ctx=None` rather than `={}` — it is the one check that writes to it, and a default argument is one dict shared by every caller that omits one.

**Not taken here: the semantics change.** Telling a solver that is *absent* from one that ran and rejected the deck needs a distinct outcome carried across the sandbox wrapper protocol — `SolverMissing` exists in the implementation, but nothing survives the crossing except a message. That is a design change rather than a fix, it is the author's call, and it is recorded on the PR for them. The caching defect is independent of it and was wrong either way, which is why it is fixed now and separately.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/partcad/shape.py [hunk 7]
```diff
@@ -1209,6 +1248,272 @@ def render(
...
+            ctx.ensure_dirs_for_file(final_filepath)
+            # A model is the answer to *this* run, and the path it goes to is
+            # stable -- '<part>.<analysis>.<extension>', beside the package. So
+            # one an earlier run left there would satisfy the check below and be
+            # handed back as the new result: last week's stresses under today's
+            # load, with nothing to say they are not today's. Removed before the
+            # implementation is asked, which makes the file's existence
+            # afterwards mean what it is read as meaning.
+            if os.path.exists(final_filepath):
+                os.remove(final_filepath)
```
[the remark this answers is not in the document]
Comment (clairbee) [reply src/partcad/shape.py]: Real, and reachable — thank you. The IDE's FEA tab issues a request per Run, so changing the implementation field and running twice puts two `analyze_async` calls on the same `<part>.<analysis>.<extension>` path with different `kwargs`. The `os.remove` I added is what sharpened it: before, the risk was a stale file being handed back; now a concurrent call can delete a model between the moment its owner wrote it and the moment the existence check reads it, so a run that succeeded reports "produced no model", or returns a path holding the other run's data.

Not fixed in this PR, and I want to be plain about why rather than wave it off.

The clean fix is for the implementation to write to a path unique to the invocation and for `analyze_async` to move it into place, which removes both the delete window and the ambiguity — the fresh path's absence unambiguously means "this run produced nothing", which is what the check is actually asking. `os.remove` then goes away entirely. Returning a *per-run* path as you suggest goes further, but the stable path is a documented part of the contract (`<part>.<analysis>.<extension>`, beside the package; the CLI, the viewer and the example README all rely on it), so changing what is returned is a design decision rather than a repair.

What stops me doing the write-and-move version now is that I cannot validate it. This container lost its virtualenv to a restart and has no Docker for the dev container, so `pytest` will not run here; the change touches the sandbox-facing path in `shape.py`, and the tests that cover it are exactly the ones I could not execute. Pushing it unvalidated on top of an otherwise green PR is the trade I would rather not make silently. Flagged for the author with the proposed shape of the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: examples/feature_cae/partcad.yaml [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      The bore of a 10x100 mm pipe, driven by 5 mN at the inlet and held no-slip
+      along the wall. Fully developed laminar flow: Hagen-Poiseuille.
+    implements:
+      pipe-inlet: [[0, 0, -50], [0, 0, 1], 0]
+      pipe-wall:
+    cfd:
+      fix:
+        - pipe-wall
+      load:
+        pipe-inlet: 0.005 N
```
[the remark this answers is not in the document]
Comment (clairbee) [reply examples/feature_cae/partcad.yaml:151]: Your facts are right and your conclusion is the one I disagree with, so let me put the disagreement rather than half-take it.

Correct on all three counts: `AnalysisConfig` takes only `fix:` and `load:`, so the outlet cannot be declared; the solve does not converge; and `pc test -f cfd` cannot validate this case, because it skips a solver exception. That last one is now sharper than when you wrote it — `c789802` stops the skip being *cached*, but it is still a pass.

Where I part company is on removing the declaration. This example exists to be the thing that says what does not work. It was written as a case with a closed-form answer precisely so a wrong answer would be visible, and it earned that in its first run: four defects, one of which returned `0.0000 mm` — a valid deck, a happy solver, and a number a reader could have believed. Deleting the `cfd:` section would delete the only artefact in the tree that demonstrates the outlet gap, and would leave `pc cae cfd` shipped, documented and defaulted to an implementation with nothing exercising it at all. The README leads with "This one does not run at all, and not for the reason predicted", names the divergence, names the missing outlet as structural, and tells the reader not to expect a number. That is the example doing its job, not a claim that it passes.

Implementing outlet support instead is the right eventual answer and is out of this PR's scope: it means a third key in `cae:`, schema and `AnalysisConfig` changes, the boundary plumbing, and the CFD implementation in a separate repository — with no way to verify any of it from here, since this machine has no CalculiX.

So: declaration kept, gap documented, and the design question raised with the author on the PR rather than settled by me. If they would rather the `cfd:` section came out until it converges, that is a one-line change and I will make it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: examples/feature_cae/cantilever.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+# A 100 x 10 x 10 mm bar, centred on the origin and lying along X, so it runs
+# from x = -50 (the clamped root) to x = +50 (the loaded tip). Centred rather
+# than aligned to a face on purpose: `Box` centres by default, and a validation
+# case is the last place to spend correctness on a nicety of placement.
+with bd.BuildPart() as result:
+    bd.Box(100, 10, 10)
+
+if "show_object" in locals():
+    show_object(result.part.wrapped, name="cantilever")
```
[the remark this answers is not in the document]
Comment (clairbee) [reply examples/feature_cae/cantilever.py:11]: Not taking this one, because it would make these two files the odd ones out.

`if "show_object" in locals():` is the convention for the optional viewer hook across this repository's examples — `examples/feature_render_custom/bracket.py:35`, `examples/produce_part_build123d_primitive/cube.py:6` and two others use exactly this, and several more (`examples/consume_build123d/basic.py:8`, `examples/consume_cadquery/basic.py:8`, `examples/consume_sdf/basic.py:8`) call `show_object` bare with no guard at all, which is a plainer F821 than this one. Rewriting two of the five leaves a reader wondering why these differ.

The report is not wrong — F821 is not in the `ignore` list under `[tool.ruff.lint]`, so `pc lint` does flag it — but it is a false positive in substance: at module scope `locals()` *is* `globals()`, so the guard does what it says and the call cannot execute undefined. The lint tool's inability to see that is the same for all five files.

Worth noting the sibling comment on `examples/feature_cae/pipe.py` proposes `locals().get(...)` where this one proposes `globals().get(...)` for the identical construct, which is itself a sign the fix is not settled.

If it should be fixed, it should be fixed as one sweep over all of `examples/` together with a decision about the unguarded calls — and that is its own change, not a drive-by in a PR about engineering analysis.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: examples/feature_cae/pipe.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+# The *fluid volume*, not the tube around it: a solid cylinder of the bore.
+# 10 mm across and 100 mm long, centred on the origin and lying along Z, so it
+# runs from z = -50 (the inlet) to z = +50 (the outlet).
+#
+# Declaring the bore rather than the casting is what `cfd:` means by a part.
+with bd.BuildPart() as result:
+    bd.Cylinder(radius=5, height=100)
+
+if "show_object" in locals():
+    show_object(result.part.wrapped, name="pipe")
```
[the remark this answers is not in the document]
Comment (clairbee) [reply examples/feature_cae/pipe.py:12]: Same answer as on `examples/feature_cae/cantilever.py:11`, and for the same reason: `if "show_object" in locals():` is the existing convention across this repository's examples (four other files use it, three more call `show_object` unguarded), so rewriting two of them makes these the inconsistent ones. At module scope `locals()` is `globals()`, so the guard is sound and the F821 is a limitation of the linter rather than a defect here.

Note that this comment proposes `locals().get(...)` while the one on `cantilever.py` proposes `globals().get(...)` for the identical construct — which is part of why I would rather this were settled once, as a sweep over all of `examples/`, than applied two different ways in one PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/partcad/shape.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (clairbee) [reply src/partcad/shape.py]: Thanks — and your restatement of the fix is exactly right: unique invocation path, then move the finished model onto the documented stable path. That keeps the contract and removes the stale-file case, the concurrent-delete case and the wrong-model case in one go.

On the follow-up issue: please don't open one from this thread. It is the maintainer's tracker and their call whether this lands as an issue, as a follow-up PR, or inside this one before merge — and it is already recorded in three places they will see (this thread, the PR conversation, and my summary to them). A bot-opened issue on top of that is one more thing for them to triage rather than one less.

Leaving the thread open is the right call and I have not resolved it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: examples/feature_cae/cantilever.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (clairbee) [reply examples/feature_cae/cantilever.py:11]: Same answer as on the `shape.py` thread: please don't open a follow-up issue from here. Whether the `show_object` sweep is worth doing — and whether the answer is a rewrite, a per-file `# noqa: F821`, or adding `F821` to the `[tool.ruff.lint]` ignores for `examples/` — is a lint-policy decision for the maintainer, and the three options are not equivalent. It is recorded on this thread and in my summary to them, which is enough for it to be found.

Agreed on the substance: the guard is runtime-safe at module scope, and any change should cover the guarded hooks, the unguarded ones and the Ruff policy together.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/partcad/shape.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (clairbee) [reply src/partcad/shape.py]: Fixed in `1846a4b`, but by the lock rather than by a unique path — and your last sentence is right, so let me be explicit about which half this closes.

**Closed.** `analyze_async` now holds the shape's *own* instantiation lock across the whole clear-run-verify sequence, so the interleavings are gone: a second run's `os.remove` can no longer land between this run's write and its existence check, and neither the check nor the returned model can belong to another run. It is the same lock `get_wrapped` already took — one lock for a shape, because the output path is derived from the shape — and it is re-entrant because the operations nest (`analyze_async` → `get_wrapped` → `_run_implementation_async`), which a bare `asyncio.Lock` would have deadlocked on. Taking it in `_run_implementation_async` means `export:` and `render:` get the same protection, not just `cae:`.

**Not closed, deliberately.** "A lock alone does not preserve the first caller's returned path after a later call overwrites it" — true, and still true. `<part>.<analysis>.<extension>` beside the package is the *shape's* model file, not this call's; a later run of the same analysis on the same shape replaces it after this one has returned. That is the contract `pc render` and `pc export` already have, and it is what lets the IDE and the CLI both name the file without being told where it went. Giving each run its own path takes that name away from everyone who relies on it, so it is a different change from this one. `e71233a` says exactly that in a comment at the lock, so the next reader does not have to work out whether it was intended.

The cost taken knowingly: two *different* outputs of one shape no longer overlap, even though they write different paths. PartCAD's parallelism is across shapes rather than within one.

Covered by `tests/partcad/unit/test_shape_lock.py` — one holder at a time, re-entrancy within a task, a second task still waiting, release on exception, two runs not taking each other's model, and per-loop locking across threads.

---
_Generated by [Claude Code](https://claude.ai/code)_
